src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -745,22 +745,64 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
     val p = PolicyTable.syntax("p")
     val g = GroupTable.syntax("g")
 
-    serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
-      // first get all the group PKs associated to all the resource policies
-      // then delete all the resource policies
-      // then delete all the groups from the first query
+    val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
 
-      val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
+    for {
+      groupPKsToDelete <- readOnlyTransaction("deletePolicyFindGroups", samRequestContext) { implicit session =>
+          samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+        }
+      writeResult <- serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
+        // first get all the group PKs associated to all the resource policies
+        // then delete all the resource policies
+        // then delete all the groups from the first query
+        samsql"""delete $fromAndWhereFragment""".update().apply()
+
+        if (groupPKsToDelete.nonEmpty) {
+          samsql"""delete from ${GroupTable as g}
+                 where ${g.id} in ($groupPKsToDelete)""".update().apply()
+        }
+      }).attempt
+      _ <- handleResult(samRequestContext, writeResult, groupPKsToDelete)
+    } yield ()
+  }
+
+  private def handleResult(samRequestContext: SamRequestContext, writeResult: Either[Throwable, AnyVal], groupPKsToDelete: List[GroupPK]) = {
+    writeResult match {
+      case Left(fkViolation: PSQLException) if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
+        handleForeignKeyViolation(samRequestContext, groupPKsToDelete)
+      case Left(e) => IO.raiseError(e)
+      case Right(_) => IO.unit
+    }
+  }
 
-      val groupPKsToDelete = samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+  private def handleForeignKeyViolation(samRequestContext: SamRequestContext, groupPKsToDelete: List[GroupPK]) = {
+    for {
+      problematicGroups <- getGroupsCausingForeignKeyViolation(samRequestContext, groupPKsToDelete)
+      _ <- IO.raiseError[Unit](new WorkbenchExceptionWithErrorReport( // throws a 500 since that's the current behavior
```

Comment:
Out of curiosity, why is `[Unit]` needed here but not when we're using `IO.raiseError` above?

Reply:
I'm not sure why, but without the `[Unit]` the compiler says `dead code following this construct`.

```
[error] /app/src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:782:9: dead code following this construct
[error]       _ <- IO.raiseError(new WorkbenchExceptionWithErrorReport( // throws a 500 since that's the current behavior
[error]         ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

I also get the same error if I put add a [Nothing]. (`IO.raiseError(e)` return type is IO[Nothing]). 

![Screen Shot 2021-09-23 at 9 29 38 AM](https://user-images.githubusercontent.com/5375000/134516194-3affdc74-ec2d-495e-a817-716e758eb2bc.png)

Seems like there's technically a type mismatch but in the rest of sam code it seems like we don't usually need to typecast it and it doesn't cause any issues.